src/main/scala/firrtl/transforms/InlineBooleanExpressions.scala

```diff
@@ -85,22 +86,34 @@ class InlineBooleanExpressions extends Transform with DependencyAPIMigration {
     var inlineCount: Int = 1
 
     /** Whether or not an can be inlined
+      * @param ref the WRef that references refExpr
       * @param refExpr the expression to check for inlining
       * @param outerExpr the parent expression of refExpr, if any
       */
-    def canInline(refExpr: Expression, outerExpr: Option[Expression]): Boolean = {
+    def canInline(ref: WRef, refExpr: Expression, outerExpr: Option[Expression]): Boolean = {
       val contextInsensitiveDetOps: Set[PrimOp] = Set(Lt, Leq, Gt, Geq, Eq, Neq, Andr, Orr, Xorr)
       outerExpr match {
         case None => true
-        case Some(o) if (o.tpe == Utils.BoolType) =>
-          refExpr match {
-            case _: Mux => false
-            case e => e.tpe == Utils.BoolType
-          }
         case Some(o) =>
-          refExpr match {
-            case DoPrim(op, _, _, Utils.BoolType) => contextInsensitiveDetOps(op)
-            case _                                => false
+          if ((refExpr.tpe != Utils.BoolType) || refExpr.isInstanceOf[Mux]) {
+            false
+          } else {
+            o match {
+              // if outer expression is also boolean context does not affect width
+              case o if o.tpe == Utils.BoolType => true
+
+              // mux condition argument is self-determined
+              case m: Mux if m.cond eq ref => true
+
+              // dshl/dshr second argument is self-determined
+              case DoPrim(Dshl | Dshr, Seq(_, shamt), _, _) if shamt eq ref => true
```

Comment:
This should probably include `Dshlw` given where this pass runs.

Reply:
You could also do a backtick match on shamt (and the mux cond) for the pattern, but that is debatably either more or less clear.